Adopt safer CPP in ApplicationManifestParser.cpp

jelee53 opened the pull request:
#### 3e9b5c7df9a9d30d267b61bfd7e345bc4869e251
<pre>
Adopt safer CPP in ApplicationManifestParser.cpp
<a href="https://bugs.webkit.org/show_bug.cgi?id=304342">https://bugs.webkit.org/show_bug.cgi?id=304342</a>
<a href="https://rdar.apple.com/166719936">rdar://166719936</a>

Reviewed by Chris Dumez.

Apply <a href="https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.">https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.</a>

No new tests needed.

* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::ApplicationManifestParser):
(WebCore::ApplicationManifestParser::parseIcons):
(WebCore::ApplicationManifestParser::parseShortcuts):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:

Canonical link: <a href="https://commits.webkit.org/304658@main">https://commits.webkit.org/304658@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/0e6cd5d2aaf0d4db1836145461301a058802d6dd

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/136186 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/8542 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/47466 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/143895 "Built successfully") | [  ~~🛠 win~~](https://ews-build.webkit.org/#/builders/59/builds/89154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6ae93b45-865f-4b69-a234-621414cdb89a) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/138058 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/9217 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/8387 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/104133 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/59/builds/89154 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/a54ec386-aacf-4e82-8fef-2ff806c2dc8d) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/139132 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/6698 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/122031 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/84961 "Passed tests") | | [✅ 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/3f4d6e79-5ab4-4444-a1fb-950f5d7cc7ea) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/6368 "Passed tests") | [⏳ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-API-Tests-EWS "Waiting to run tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/4490 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/115651 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/40229 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/146641 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/8226 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/40798 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/112478 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/8242 "Built successfully") | [⏳ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/macOS-Tahoe-Debug-WK2-Tests-EWS "Waiting to run tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/112814 "Passed tests") | | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/6289 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/118335 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/62213 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/20985 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/8274 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/36393 "Passed tests") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/7992 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/71833 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/8213 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/8066 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -121,8 +121,8 @@ ApplicationManifest ApplicationManifestParser::parseManifest(const JSON::Object&
     parsedManifest.orientation = parseOrientation(manifest);
 
-    if (m_document)
-        m_document->processApplicationManifest(parsedManifest);
```
I think we can just mark `m_document` as `const` instead.

cdumez commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -320,9 +320,9 @@ Vector<ApplicationManifest::Icon> ApplicationManifestParser::parseIcons(const JS
         if (!iconObject)
             continue;
-        const auto& iconJSON = *iconObject;
+        const Ref iconJSON = *iconObject;
```
We can drop the `const`.

cdumez commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -399,9 +399,9 @@ Vector<ApplicationManifest::Shortcut> ApplicationManifestParser::parseShortcuts(
         if (!shortcutObject)
             continue;
-        const auto& shortcutJSON = *shortcutObject;
+        const Ref shortcutJSON = *shortcutObject;
```
We can drop the `const`.

cdumez commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
```diff
@@ -67,7 +67,7 @@ class ApplicationManifestParser {
     void logDeveloperWarning(const String& message);
 
-    RefPtr<Document> m_document;
+    const RefPtr<Document> m_document;
```
While you're modifying this file, do you mind modifying the constructor and use `RefPtr<Document>&&` for the parameter instead of `RefPtr<Document>` and then use `WTFMove()` internally?

cdumez reviewed (approved):
LGTM with nit.

jelee53 committed:
Adopt safer CPP in ApplicationManifestParser.cpp
https://bugs.webkit.org/show_bug.cgi?id=304342
rdar://166719936

Reviewed by Chris Dumez.

Apply https://github.com/WebKit/WebKit/wiki/Safer-CPP-Guidelines.

No new tests needed.

* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
(WebCore::ApplicationManifestParser::ApplicationManifestParser):
(WebCore::ApplicationManifestParser::parseIcons):
(WebCore::ApplicationManifestParser::parseShortcuts):
* Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.h:
* Source/WebCore/SaferCPPExpectations/UncheckedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedCallArgsCheckerExpectations:
* Source/WebCore/SaferCPPExpectations/UncountedLocalVarsCheckerExpectations:

Canonical link: https://commits.webkit.org/304658@main

annevk commented on Source/WebCore/Modules/applicationmanifest/ApplicationManifestParser.cpp:
```diff
@@ -320,9 +320,9 @@ Vector<ApplicationManifest::Icon> ApplicationManifestParser::parseIcons(const JS
         if (!iconObject)
             continue;
-        const auto& iconJSON = *iconObject;
+        Ref iconJSON = *iconObject;
```
FWIW, I think we could just do `RefPtr iconObject` above and disregard this whole line. Same below with `shortcutObject`.

cdumez replied:
I agree this would have been nicer. I missed that.
